### DISCUSSION TITLE
upgraded to 1.94.1 and enabled “Automatic library watching (experimental)”,error:ENOSPC: System limit for number of file watchers reached

### DISCUSSION THREAD

**Discussion by nodis** [immich-app/immich#6833] [Q&A] [answered]
### The bug

Hello, I upgraded to 1.94.1 and enabled “Automatic library watching (experimental)”, but the following error occurred:
```
[Nest] 7 - 02/01/2024, 3:51:28 AM ERROR [LibraryService] Library watcher for library c69faf55-f96d-4aa0-b83b-2d80cbc27d98 encountered error: Error: ENOSPC: System limit for number of file watchers reached, watch '/media/volume2/homes/xxx/Photos/MobileBackup/Pro/DCIM/Screenshots/@eaDir/Screenshot_2015-12-25-23-41-06_com.excelliance.dualaid.png'
2024/02/0111:51:28
[Nest] 7 - 02/01/2024, 3:51:28 AM ERROR [LibraryService] Library watcher for library c69faf55-f96d-4aa0-b83b-2d80cbc27d98 encountered error: Error: ENOSPC: System limit for number of file watchers reached, watch '/media/volume2/homes/xxx/Photos/MobileBackup/Pro/DCIM/Screenshots/@eaDir/Screenshot_2015-12-25-23-41-08_com.excelliance.dualaid.jpg'
2024/02/0111:51:28
[Nest] 7 - 02/01/2024, 3:51:28 AM ERROR [LibraryService] Library watcher for library c69faf55-f96d-4aa0-b83b-2d80cbc27d98 encountered error: Error: ENOSPC: System limit for number of file watchers reached, watch '/media/volume2/homes/xxx/Photos/MobileBackup/Pro/DCIM/Screenshots/@eaDir/Screenshot_2015-12-25-23-41-06_com_1457834466869.excelliance.dualaid.png'
2024/02/0111:51:28
[Nest] 7 - 02/01/2024, 3:51:28 AM ERROR [LibraryService] Library watcher for library c69faf55-f96d-4aa0-b83b-2d80cbc27d98 encountered error: Error: ENOSPC: System limit for number of file watchers reached, watch '/media/volume2/homes/xxx/Photos/MobileBackup/Pro/DCIM/Screenshots/@eaDir/Screenshot_2015-12-25-23-41-08_com_1457834467052.excelliance.dualaid.png'

``` 

### The OS that Immich Server is running on

Synology

### Version of Immich Server

1.94.1

### Version of Immich Mobile App

1.94.1

### Platform with the issue

- [X] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
xx
```

### Your .env content

```Shell
xx
```

### Reproduction steps

```bash
1.
2.
3.
...
```

**Comment by bo0tzz** [comment]
cc @etnoy

**Comment by etnoy** [comment]
You need to increase the linux limit for the number of watchers.

In sysctl, change the following value to a suitable number: fs.inotify.max_user_watches=524288

https://dev.to/shawon/fix-error-enospc-system-limit-for-number-of-file-watchers-reached-pfh

**Reply by cr0n76** [reply]
I encountered the same issue. Even setting fs.inotify.max_user_watches to 1024*1024 does not solve the problem.
The main issue is, that the server UI is not responding, though. So I have no possibility to deactivate the option again. 

Do you have any suggestions to deactivate it without access to the UI?

**Reply by cr0n76** [reply, marked as answer]
A short update on how to deactivate it without access to the UI:
If the [config file](https://immich.app/docs/install/config-file/) is not in use, the config is stored in the postgres database. 
Check [here](https://immich.app/docs/guides/database-queries/) on how to query it.

To update the config to disable automatic library watching one can use 
`UPDATE system_config SET value = false WHERE key = 'library.watch.enabled';`  

However the neccessity to update any config value by this should be very rare. Don't do this unless you are really sure that this is the only way and is a sane decision.

**Reply by iFloris** [reply]
Thanks for figuring this out, I was stuck like you

**Reply by etnoy** [reply]
Please file a feature request for disabling watching via external means, I'll try to create that

**Reply by cr0n76** [reply]
To be honest I would rather file a request to delete the option until it does not bear the possibility to render the UI unresponsive anymore.

On the other hand the word "EXPERIMENTAL" is clearly stated! And the fix was pretty straight forward. 
Maybe to add a short info that this may happen (and to solve it via postgres or the config file) is the better thing to do right now.

But for me it does not feel clean to implement a new "feature" that is only needed to react to a bug from another feature, which is clearly stated as experimental.

**Reply by dharapvj** [reply]
in immich 1.116.0 - I had to use this..

```sql
# First select and get the value.
select * from system_metadata where key = 'system-config';

# now update .. but copy your system-config value from above.
BEGIN;
update system_metadata
set value= '{"job": {"smartSearch": {"concurrency": 1}, "faceDetection": {"concurrency": 4}}, "library": {"watch": {"enabled": false}}, "metadata": {"faces": {"import": true}}}'
where key = 'system-config';

# verify
select * from system_metadata where key = 'system-config';
COMMIT;

```

**Comment by iansewell** [comment]
thanks! I was stuck as well on this and this solved my problem, thank you @cr0n76

**Comment by liviumac** [comment]
Hi, I updated today and I can no longer access the interface.
I am seeing a lot of error like the one below

[Nest] 7  - 10/15/2024, 8:09:27 PM   ERROR [Microservices:LibraryService] Library watcher for library 463fc65b-c378-4407-a6ae-7decd6dfb58f encountered error: Error: ENOSPC: System limit for number of file watchers reached, watch '/srv/dev-disk-by-uuid-9bda9ca2-0860-45b2-92f7-bec289ccb9aa/Seagate4T/Poze/Dropbox/PC Diverse/Promotiematerialen en templates/Styleguide EU/Product Images/New Packaging/Groups/Skin Recovery/Rosacea_WithoutLotion.png'

Can someone guide through solving this? I am a complete noob. I do not even know where to run the code given above.

I am using portainer on an OMV server. Do I use putty for the above script or what? I tried running it in the container console but it did not work.

Thank you

PS "In rare cases, the library watcher can hang, preventing Immich from starting up. In this case, disable the library watcher in the configuration file. If the watcher is enabled from within Immich, the app must be started without the microservices. Disable the microservices in the docker compose file, start Immich, disable the library watcher in the admin settings, close Immich, re-enable the microservices, and then Immich can be started normally."

there is no microservices in the compose file (anymore?)

**Reply by connor-tan** [reply]
i got a same problem, have you been solve this issue?

**Reply by Qrrbrbirlbel** [reply]
Indeed, I don't see a microservices in my compose file either. However, the [manual](https://immich.app/docs/developer/architecture/) still states

> Not shown in the diagram, is that the server is split into two separate containers `immich-server` and `immich-microservices`. The microservices container does not handle API requests or schedule cron jobs, but primarily handles incoming job requests from Redis.

and

> ## Server
>
> The Immich backend is divided into several services, which are run as individual docker containers.
>
>    * `immich-server` - Handle and respond to REST API requests
>    * `immich-microservices` - Execute background jobs (thumbnail generation, metadata extraction, transcoding, etc.)
>    * `immich-machine-learning` - Execute machine learning models
>    * `postgres `- Persistent data storage
>    * `redis`- Queue management for `immich-microservices`

**Reply by DSchmidtDev** [reply]
@Qrrbrbirlbel in the administration docs it's explained how to split microservice and server in different containers in docker compose
https://immich.app/docs/administration/jobs-workers#split-workers

**Comment by mkaflowski** [comment]
I’m wondering if setting watches for specific external libraries would be a good idea—or perhaps adding a “filter folders” option would be even better, to avoid configuring libraries from scratch. For example, if you have an old photo folder, there's no need to watch it all the time.
